api-guidelines/v4.0/API-DESIGN-GUIDELINES/DISCOVERY-API.md

```diff
@@ -0,0 +1,170 @@
+application. This will always include the core Ed-Fi (Unifying) Data Model.
+Other models listed are the _extensions_ that a given deployment implements. The
+example above shows the core Teacher Preparation Data Model (TPDM) extension.
+
+## URLs Section
+
+The URLs section helps client applications discover for themselves how to
+interact with the Ed-Fi API application. The _required_ URLs are further
+described in the following sections.
+
+The optional values listed below represent features of the the Ed-Fi ODS/API
+Platform. Other applications can add their own URLs as needed.
+
+```json
+"xsdMetadata": "https://api.ed-fi.org/v7.0/api/metadata/xsd",
+"changeQueries": "https://api.ed-fi.org/v7.0/api/changeQueries/v1/",
+"composites": "https://api.ed-fi.org/v7.0/api/composites/v1/",
+"identity": "https://api.ed-fi.org/v7.0/api/identity/v2/"
+```
+
+### Dependencies
+
+All implementations of an Ed-Fi API _should_ expose an endpoint for dependency
+metadata.  This endpoint will provide developers with the data needed to load
+resources in the appropriate order. The endpoint _should_ have a JSON
+implementation and optional a GraphML implementation.  A sample of the JSON
+output is below:
+
+```json
+[
+  {
+    "resource": "/ed-fi/absenceEventCategoryDescriptors",
+    "order": 1,
+    "operations": [
+      "Create"
+    ]
+  },
+  {
+    "resource": "/ed-fi/academicHonorCategoryDescriptors",
+    "order": 1,
+    "operations": [
+      "Create"
+    ]
+  },
+  …
+  {
+    "resource": "/tpdm/surveySectionResponsePersonTargetAssociations",
+    "order": 22,
+    "operations": [
+      "Create"
+    ]
+  }
+]
+```
+
+### OpenApiMetadata
+
+An Ed-Fi API implementation _should_ declare itself in OpenAPI; the version of
+OpenAPI is up to the implementation. The OpenAPI specification must be a
+faithful representation of all available resources supported by the API. A given
+API application might support multiple specifications, for example one for the
+core Ed-Fi data model resources, and one for Ed-Fi descriptors. The
+specifications provided by an API application _must_ be consistent with the
+official [Ed-Fi API Specifications](../../../api-specifications/).
+
+### OAuth
+
+The OAuth URL must be the token URL used by clients for authentication. The URL
+is not required to be on the same base address, i.e. it could be hosted on
+another server.
+
+### DataManagementApi
+
+This URL represents the base path for constructing the full [Uniform Resource
+Locators](./UNIFORM-RESOURCE-LOCATORS.md) (URL) to a given resource. To this
+base path, a client appends the data model namespace. For example, the
+`/ed-fi/students` resource can be reached at the URL formed by combining the
+`dataManagementApi` URL with the resource information. From the example above,
+this yields the URL `https://api.ed-fi.org/v7.1/api/data/v3/ed-fi/students`.
+While the Ed-Fi ODS/API Platform uses "data/v3" in the URL, any implementation
+could use different path segments.
```

Comment:
known issue maybe worth at least noting: Most state implementations I know of have a single unchanging URL exposing the Discovery API. However, part of the "as-needed-path-segments" includes a school year meaning they really have  multiple dataManagementAPI paths `.../data/v3/2023/students`, `.../data/v3/2024/students`. I suspect they just hide the Discovery API for previous years, but it is mildly confusing and at minimum something I want you to be aware of.

Reply:
Glad you called this out. ODS/API 7.x handles this a _little_ better by having a placeholder in the URL: `{tenantIdentifier}/data/v3/` instead of hard-coding a single year or a tenant name. But a client application needs to know what tenant identifier to insert into that URL, which is unfortunate. I'll talk with Vinaya about this. I think we could have done something better. It would be better if we could give you a tenant-specific discovery API endpoint, and then the Discovery API would not need a `{tenantIdentifier}` placeholder.

Need to think carefully about how to represent this in the documentation, and I'll create a ticket seeing if we can make an improvement in ODS/API 7.3 this fall.

Reply:
I'm relieved by this: I just tested ODS/API 7.1 on an internal staging server and I see that the discovery API is available from the tenant-specific URL. Thus although the root-level document has `{tenantIdentifier}` place holders, the host should be giving the vendor the specific URL, e.g. `https://api-stage.ed-fi.org/Multitenant_v7.2/api/tenant1`, and this is itself a root level Discovery document whose URLs _do not_ include placeholders.

Still, I should add a note about legacy ODS/API year specific mode.